src/order/interval.lean

```diff
@@ -0,0 +1,422 @@
+      exact ⟨sup_le hb.1 hc.1, le_inf hb.2 hc.2⟩,
+      exact ⟨hb.1.trans $ a.fst_le_snd.trans hc.2, hc.1.trans $ a.fst_le_snd.trans hb.2⟩,
+    end
+  end,
+  ..interval.semilattice_sup }
+
+@[simp] lemma to_set_inf (a b : interval α) : (a ⊓ b).to_set = a.to_set ∩ b.to_set :=
+begin
+  cases a,
+  { rw [with_bot.none_eq_bot, bot_inf_eq],
+    exact (empty_inter _).symm },
+  cases b,
+  { rw [with_bot.none_eq_bot, inf_bot_eq],
+    exact (inter_empty _).symm },
+  refine (_ : to_set (dite _ _ _) = _).trans Icc_inter_Icc.symm,
+  split_ifs,
+  { refl },
+  { exact (Icc_eq_empty $ λ H,
+      h ⟨le_sup_left.trans $ H.trans inf_le_right, le_sup_right.trans $ H.trans inf_le_left⟩).symm }
+end
+
+@[simp] lemma disjoint_to_set (a b : interval α) : disjoint a.to_set b.to_set ↔ disjoint a b :=
+by { rw [disjoint, disjoint, ←(@to_set α _).le_iff_le, to_set_inf], refl }
+
+end lattice
+
+section complete_lattice
+variables [complete_lattice α]
+
+noncomputable instance : complete_lattice (interval α) :=
+by classical; exact
+{ Sup := λ s, if h : s ⊆ {⊥} then ⊥ else some
+    ⟨⟨⨅ (a : nonempty_interval α) (h : ↑a ∈ s), a.fst,
+      ⨆ (a : nonempty_interval α) (h : ↑a ∈ s), a.snd⟩, begin
+        obtain ⟨a, hs, ha⟩ := not_subset.1 h,
+        lift a to nonempty_interval α using ha,
+        exact infi₂_le_of_le a hs (le_supr₂_of_le a hs a.fst_le_snd)
+      end⟩,
+  le_Sup := λ s a ha, begin
+    split_ifs,
+    { exact (h ha).le },
+    cases a,
+    { exact bot_le },
+    { exact with_bot.some_le_some.2 ⟨infi₂_le _ ha, le_supr₂_of_le _ ha le_rfl⟩ }
+  end,
+  Sup_le := λ s a ha, begin
+    split_ifs,
+    { exact bot_le },
+    obtain ⟨b, hs, hb⟩ := not_subset.1 h,
+    lift a to nonempty_interval α using ne_bot_of_le_ne_bot hb (ha _ hs),
+    exact with_bot.coe_le_coe.2 ⟨le_infi₂ $ λ c hc, (with_bot.coe_le_coe.1 $ ha _ hc).1,
+      supr₂_le $ λ c hc, (with_bot.coe_le_coe.1 $ ha _ hc).2⟩,
+  end,
+  Inf := λ s, if h : ⊥ ∉ s ∧ ∀ ⦃a : nonempty_interval α⦄, ↑a ∈ s → ∀ ⦃b : nonempty_interval α⦄,
+    ↑b ∈ s → a.fst ≤ b.snd then some
+      ⟨⟨⨆ (a : nonempty_interval α) (h : ↑a ∈ s), a.fst,
+        ⨅ (a : nonempty_interval α) (h : ↑a ∈ s), a.snd⟩,
+          supr₂_le $ λ a ha, le_infi₂ $ h.2 ha⟩ else ⊥,
+  Inf_le := λ s a ha, begin
+    split_ifs,
+    { lift a to nonempty_interval α using ne_of_mem_of_not_mem ha h.1,
+      exact with_bot.coe_le_coe.2 ⟨le_supr₂ a ha, infi₂_le a ha⟩ },
+    { exact bot_le }
+  end,
+  le_Inf := λ s a ha, begin
+    cases a,
+    { exact bot_le },
+    split_ifs,
+    { exact with_bot.some_le_some.2 ⟨supr₂_le $ λ b hb, (with_bot.coe_le_coe.1 $ ha _ hb).1,
+        le_infi₂ $ λ b hb, (with_bot.coe_le_coe.1 $ ha _ hb).2⟩ },
+    rw [not_and_distrib, not_not] at h,
+    cases h,
+    { exact ha _ h },
+    cases h (λ b hb c hc, (with_bot.coe_le_coe.1 $ ha _ hb).1.trans $ a.fst_le_snd.trans
+      (with_bot.coe_le_coe.1 $ ha _ hc).2),
+  end,
+  ..interval.lattice, ..interval.bounded_order }
```

Comment:
Should I take in `decidable_rel (≤)` as an assumption to avoid the derived `lattice (interval α)` being poisoned by `classical`?

Reply:
Definitely

Reply:
Wait, what derived instance?

Reply:
Ah, I see; yes you need to do that to avoid an instance diamond